contracts/src/lib/LibGateway.sol

```diff
@@ -416,6 +416,100 @@ library LibGateway {
+            });
+            return (true, applyRet);
+        }
+
+        // The first thing we do is to find out the directionality of this message and act accordingly,
+        // incrasing the applied nonces conveniently.
+        // slither-disable-next-line uninitialized-local
+        SupplySource memory supplySource;
+        IPCMsgType applyType = crossMsg.applyType(s.networkName);
+        if (applyType == IPCMsgType.BottomUp) {
+            // Load the subnet this message is coming from. Ensure that it exists and that the nonce expectation is met.
+            (bool registered, Subnet storage subnet) = LibGateway.getSubnet(arrivingFrom);
+            if (!registered) {
+                // this means the subnet that sent the bottom up message is not registered,
+                // we cannot send the receipt back as top down because the subnet is not registered
+                // we ignore this message for as it's not valid, and it may be someone trying to forge it.
+                return (false, applyRet);
+            }
+            if (subnet.appliedBottomUpNonce != crossMsg.nonce) {
+                return (true, ApplyMsgRet({
+                    crossMsg: crossMsg,
+                    success: false,
+                    ret: abi.encodeWithSelector(InvalidCrossMsgNonce.selector)
+                }));
+            }
+            subnet.appliedBottomUpNonce += 1;
+
+            // The value carried in bottom-up messages needs to be treated according to the supply source
+            // configuration of the subnet.
+            supplySource = SubnetActorGetterFacet(subnet.id.getActor()).supplySource();
+        } else if (applyType == IPCMsgType.TopDown) {
+            // Note: there is no need to load the subnet, as a top-down application means that _we_ are the subnet.
+            if (s.appliedTopDownNonce != crossMsg.nonce) {
+                return (true, ApplyMsgRet({
+                    crossMsg: crossMsg,
+                    success: false,
+                    ret: abi.encodeWithSelector(InvalidCrossMsgNonce.selector)
+                }));
+            }
+            s.appliedTopDownNonce += 1;
+
+            // The value carried in top-down messages locally maps to the native coin, so we pass over the
+            // native supply source.
+            supplySource = SupplySourceHelper.native();
+        }
+
+        // If the crossnet destination is NOT the current network (network where the gateway is running),
+        // we add it to the postbox for further propagation.
+        // Even if we send for propagation, the execution of every message
+        // should increase the appliedNonce to allow the execution of the next message
+        // of the batch (this is way we have this after the nonce logic).
+        if (!crossMsg.to.subnetId.equals(s.networkName)) {
+            bytes32 cid = crossMsg.toHash();
+            s.postbox[cid] = crossMsg;
```

Comment:
How do you know what to pick up from the postbox later?

Reply:
good call, I copied the previous implementation, probably this is a separate issue.